include/pybind11/pytypes.h

```diff
@@ -246,6 +246,11 @@ class handle : public detail::object_api<handle> {
     const handle &inc_ref() const & {
 #ifdef PYBIND11_HANDLE_REF_DEBUG
         inc_ref_counter(1);
+#endif
+#if defined(PYBIND11_ASSERT_GIL_HELD_INCREF_DECREF)
+        if (m_ptr != nullptr && !PyGILState_Check()) {
+            throw std::runtime_error("pybind11::handle::inc_ref() PyGILState_Check() failure.");
```

Comment:
Is there a reason no pybind11_fail? Because of the corresponding `!PyErrOccurred() ` assertion?

Reply:
Yes.
But even if that was taken out in `pybind11_fail()`, I think a "raw" C++ exception is best here, we want the process to terminate as directly as possible. I was thinking of using `std::terminate()` directly, but then it is troublesome to also produce the message, so I decided simpler is better. In our environment at least, `std::runtime_error()` works out very nicely (useful traceback with message).

Reply:
FYI, functions that can raise exceptions do come with some subtle costs in C++. The compiler needs to keep track of the many ways in which might have to destroy objects if an exception arises at various different points of the program and generate code for each one. I don't think it's a performance issue, but this will likely make the debug binaries bigger (compared to `pybind11_fail`) given that such a central operation is affected.

Reply:
Narrow focus:

Currently `pybind11_fail()` includes this:
```
assert(!PyErr_Occurred());
```
That needs the GIL, but here (`handle`) we just determined that we don't have it.

Bigger picture:

There is a general tradeoff: a faster or leaner binary today vs a future with more features.

Boundary condition: human time is ~constant.

So ultimately we always trade one for the other.

If you hire person A who likes to optimize, you may get something that runs in 1.0 time units with 1.0 resources, with 10 features.

If you hire person B, who's more like me, you get something that runs in 1.1 time units with 1.1 resources, with 20 features.

What's "better"?

It depends.